pinot-core/src/main/java/com/linkedin/pinot/core/operator/docidsets/OrBlockDocIdSet.java

```diff
@@ -22,129 +22,102 @@
-          allBitmaps.add(childBitmap);
+          bitmap.or((ImmutableRoaringBitmap) docIdSet.getRaw());
         } else {
-          BlockDocIdIterator iterator = docIdSet.iterator();
-          rawIterators.add(iterator);
+          iterators.add(docIdSet.iterator());
         }
       }
-      MutableRoaringBitmap answer = allBitmaps.get(0).toMutableRoaringBitmap();
-      for (int i = 1; i < allBitmaps.size(); i++) {
-        answer.or(allBitmaps.get(i));
+      IntIterator intIterator = bitmap.getIntIterator();
+      BitmapDocIdIterator bitmapDocIdIterator = new BitmapDocIdIterator(intIterator);
+      bitmapDocIdIterator.setStartDocId(_minDocId);
+      bitmapDocIdIterator.setEndDocId(_maxDocId);
+      if (iterators.isEmpty()) {
+        return bitmapDocIdIterator;
+      } else {
+        iterators.add(bitmapDocIdIterator);
       }
-      intIterator = answer.getIntIterator();
-      BitmapDocIdIterator singleBitmapBlockIdIterator = new BitmapDocIdIterator(intIterator);
-      singleBitmapBlockIdIterator.setStartDocId(minDocId);
-      singleBitmapBlockIdIterator.setEndDocId(maxDocId);
-      rawIterators.add(singleBitmapBlockIdIterator);
-      docIdIterators = new BlockDocIdIterator[rawIterators.size()];
-      rawIterators.toArray(docIdIterators);
+      return new OrDocIdIterator(iterators.toArray(new BlockDocIdIterator[iterators.size()]), _minDocId, _maxDocId);
     } else {
-      docIdIterators = new BlockDocIdIterator[docIdSets.size()];
-      for (int srcId = 0; srcId < docIdSets.size(); srcId++) {
-        docIdIterators[srcId] = docIdSets.get(srcId).iterator();
+      int numDocIdSets = _docIdSets.size();
+      BlockDocIdIterator[] iterators = new BlockDocIdIterator[numDocIdSets];
+      for (int i = 0; i < numDocIdSets; i++) {
+        iterators[i] = _docIdSets.get(i).iterator();
       }
+      return new OrDocIdIterator(iterators, _minDocId, _maxDocId);
     }
-//    if (docIdIterators.length == 1) {
-//      return docIdIterators[0];
-//    } else {
-      OrDocIdIterator orDocIdIterator = new OrDocIdIterator(docIdIterators);
-      orDocIdIterator.setStartDocId(minDocId);
-      orDocIdIterator.setEndDocId(maxDocId);
-      return orDocIdIterator;
-//    }
   }
 
-  @SuppressWarnings("unchecked")
   @Override
   public <T> T getRaw() {
-    return (T) this.docIdSets;
-  }
-
-  @Override
-  public void setStartDocId(int startDocId) {
-    minDocId = Math.min(minDocId, startDocId);
-    updateMinMaxRange();
-  }
-
-  @Override
-  public void setEndDocId(int endDocId) {
-    maxDocId = Math.max(maxDocId, endDocId);
-    updateMinMaxRange();
-  }
-
-  @Override
-  public long getNumEntriesScannedInFilter() {
-    long numEntriesScannedInFilter = 0L;
-    for (FilterBlockDocIdSet docIdSet : docIdSets) {
-      numEntriesScannedInFilter += docIdSet.getNumEntriesScannedInFilter();
-    }
-    return numEntriesScannedInFilter;
+    throw new UnsupportedOperationException();
```

Comment:
why is this throwing exception. How will the parent operator do bitmap intersection if we dont return raw bitmap

Reply:
Only BitmapDocIdSet and SortedDocIdSet should implement this method